### PR TITLE
Elberger.maze

### PR DESCRIPTION
**Pull request by relberger**
Maze seems to be initializing with the incorrect values. Possible additional error in searchPath method.

### CODE REVIEW

File: src/main/java/elberger/maze/Maze.java [hunk 1]
```diff
@@ -4,145 +4,147 @@
...
-			neighbors.push(westNeighbor);
-			neighbors.push(southNeighbor);
+			if (!nextCell.isVisited())
+			{
+				nextCell.setVisited(true);
+				breakWalls(currentCell, nextCell);
+				stack.push(nextCell);
+			} else
+			{
+				stack.pop();
```
Comment (andrewoid) [inline src/main/java/elberger/maze/Maze.java]: if this neighbor is visited then want to check all the other neighbors before popping the stack

File: src/main/java/elberger/maze/Maze.java [hunk 2]
```diff
@@ -4,145 +4,147 @@
...
+				searchPath(nextCell);
```
Comment (andrewoid) [inline src/main/java/elberger/maze/Maze.java]: If you are using a Stack, don't use recursion

File: src/main/java/elberger/maze/Maze.java [hunk 3]
```diff
@@ -114,6 +90,35 @@ else if (currentCell.getX() == nextCell.getX() - 1)
...
+			Cell currentCell = stack.pop();
+			Cell nextCell = getNeighbors(currentCell).get(0);
+
+			if (!nextCell.isVisited())
+			{
+				nextCell.setVisited(true);
+				breakWalls(currentCell, nextCell);
+				stack.push(nextCell);
+			}
+			else if (nextCell.isVisited())
```
Comment (andrewoid) [inline src/main/java/elberger/maze/Maze.java]: by definition if the first if statement was false, then this if statement would be true

File: src/main/java/elberger/maze/Maze.java [hunk 4]
```diff
@@ -114,6 +90,35 @@ else if (currentCell.getX() == nextCell.getX() - 1)
...
+			{
+				for (int i = 0; i < getNeighbors(nextCell).size(); i++)
```
Comment (andrewoid) [inline src/main/java/elberger/maze/Maze.java]: You don't want to loop through the neighbors of the nextCell, you want to loop through the neighbors of the currentCell

**Review by relberger — COMMENTED** [review]
1) Somewhere in the for loop of the maze constructor, the values of the booleans of the maze cells get changed, as if calling other methods which I can't see as being called. 
2) If I comment out the else of searchPath(), the debugger runs, but it gets stuck at "Collecting data..." if it's included.

File: src/main/java/elberger/maze/Maze.java (hunk 4 above)
Comment (andrewoid) [reply src/main/java/elberger/maze/Maze.java]: You never addressed this comment.

File: src/main/java/elberger/maze/Maze.java [hunk 5]
```diff
@@ -4,145 +4,150 @@
...
 			}
 		}
 	}
 
-	public void generatePath()
+	@Override
+	public String toString()
 	{
-		searchPath(startingCell());
+		searchPath(maze[0][0]);
```
Comment (andrewoid) [inline src/main/java/elberger/maze/Maze.java]: Do not put this in the toString()

**Comment by relberger** [comment]
It's still stuck at "Collecting Data..." when it hits searchPath()

File: src/main/java/elberger/maze/Maze.java [hunk 6]
```diff
@@ -4,145 +4,152 @@
...
-				stack.push(nextCell);
+				nextCell.setVisited(true);
+				breakWalls(currentCell, nextCell);
+				stack.addAll(getNeighbors(currentCell));
 			}
 			else
 			{
-				stack.pop();
-				searchPath(nextCell);
+				for (int i = 0; i < getNeighbors(currentCell).size(); i++)
```
Comment (andrewoid) [inline src/main/java/elberger/maze/Maze.java]: You are calling `getNeighbors()` here 3 times.
What you've forgotten is that getNeighbors() returns a List that is SHUFFLED and be different each time it's called.

File: src/main/java/elberger/maze/Maze.java [hunk 7]
```diff
@@ -4,145 +4,152 @@
...
+			Cell nextCell = getNeighbors(currentCell).get(0);
+
+			if (!nextCell.isVisited())
 			{
-				visited.push(nextCell);
-				maze[nextCell.getX()][nextCell.getY()] = ".";
-				stack.push(nextCell);
+				nextCell.setVisited(true);
+				breakWalls(currentCell, nextCell);
+				stack.addAll(getNeighbors(currentCell));
```
Comment (andrewoid) [inline src/main/java/elberger/maze/Maze.java]: don't add the neighbors, just the nextCell

File: src/main/java/elberger/maze/Maze.java [hunk 8]
```diff
@@ -4,145 +4,152 @@
...
-		visited.push(cell);
+		cell.setVisited(true);
 
 		while (!stack.isEmpty())
 		{
-			Cell currentCell = (Cell) stack.peek();
-			Cell nextCell = getNeighbors(currentCell).getFirst();
-			if (!visited.contains(nextCell))
+			Cell currentCell = stack.pop();
+			Cell nextCell = getNeighbors(currentCell).get(0);
```
Comment (andrewoid) [inline src/main/java/elberger/maze/Maze.java]: you only care about neighbors that haven't been visited yet, not all the neighbors

File: src/main/java/elberger/maze/Maze.java [hunk 9]
```diff
@@ -4,145 +4,152 @@
...
 			}
 			else
 			{
-				stack.pop();
-				searchPath(nextCell);
+				for (int i = 0; i < getNeighbors(currentCell).size(); i++)
+				{
+					if (getNeighbors(currentCell).get(i).isVisited())
+					{
+						stack.remove(getNeighbors(currentCell).get(i));
```
Comment (andrewoid) [inline src/main/java/elberger/maze/Maze.java]: Why are you removing cells from the stack?

**Review by andrewoid — COMMENTED** [review]
You've got a lot of issues with `searchPath`.
I would start that method over again and make sure you understand how the algorithm works.

**Comment by relberger** [comment]
I changed searchPath and getNeighbors to getUnvisitedNeighbors, but now it's giving me an index out of bounds exception when it's trying to find unvisited neighbors because there are cells that are marked as visited when I'm not sure why they would be. 

Does my searchPath theoretically make sense now (other than the fact that it doesn't actually work)?

File: src/main/java/elberger/maze/MazeFromClass.java [hunk 10]
```diff
@@ -0,0 +1,113 @@
+package elberger.maze;
+
+import com.google.common.annotations.VisibleForTesting;
+
+import javax.annotation.Nullable;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+
+public class MazeFromClass
```
Comment (andrewoid) [inline src/main/java/elberger/maze/MazeFromClass.java]: Put all classes "From Class" into their own package.

File: src/main/java/elberger/maze/MazeView.java [hunk 11]
```diff
@@ -1,12 +1,105 @@
...
+			}
+			if (key.getKeyCode() == KeyEvent.VK_RIGHT && player.getX() != maze.getWidth())
+			{
+				goRight();
+			}
+		}
+
+		private void goUp()
+		{
+			if (!maze.getMaze()[player.getX() / (getWidth() / maze.getWidth())][player.getY() / (getHeight() / maze.getHeight())].isNorthWall())
```
Comment (andrewoid) [inline src/main/java/elberger/maze/MazeView.java]: Don't call the same methods over and over again.

**Pull request merged**